IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m

```diff
@@ -56,432 +62,654 @@ - (instancetype)initWithDataSource:(id<MSIDTokenCacheDataSource>)dataSource
+                    context:(id<MSIDRequestContext>)context
+                      error:(NSError **)error
+{
     if (!account.userIdentifier)
     {
         if (error)
         {
-            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInvalidInternalParameter, @"user identifier is needed to save access token for MSDIDefaultTokenCacheFormat", nil, nil, nil, context.correlationId, nil);
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInvalidInternalParameter, @"User identifier is expected for MSDIDefaultTokenCacheFormat", nil, nil, nil, context.correlationId, nil);
         }
         return NO;
     }
     
-    token.authority = [[MSIDAadAuthorityCache sharedInstance] cacheUrlForAuthority:token.authority context:context];
-    NSArray<NSURL *> *aliases = [[MSIDAadAuthorityCache sharedInstance] cacheAliasesForAuthority:token.authority];
-    
-    // delete all cache entries with intersecting scopes
-    // this should not happen but we have this as a safe guard against multiple matches
-    NSArray<MSIDToken *> *allTokens = [self getAllATsForContext:context error:error];
+    return YES;
+}
+
+#pragma mark - MSIDSharedCacheAccessor
 
-    if (!allTokens)
+- (BOOL)saveToken:(MSIDBaseToken *)token
+          account:(MSIDAccount *)account
+          context:(id<MSIDRequestContext>)context
+            error:(NSError **)error
+{
+    if (![self checkUserIdentifier:account context:context error:error])
     {
         return NO;
     }
     
-    for (MSIDToken *tokenInCache in allTokens)
+    if (token.tokenType == MSIDTokenTypeAccessToken)
     {
-        if ([tokenInCache.clientInfo.userIdentifier isEqualToString:account.userIdentifier]
-            && [tokenInCache.clientId isEqualToString:token.clientId]
-            && [tokenInCache.authority msidIsEquivalentWithAnyAlias:aliases]
-            && [tokenInCache.scopes intersectsOrderedSet:token.scopes])
+        // Access tokens have a special handling because of scopes
+        if (![self deleteAllAccessTokensWithIntersectingScopes:(MSIDAccessToken *)token
+                                                       account:account
+                                                       context:context
+                                                         error:error])
         {
-            MSIDTokenCacheKey *keyToDelete = [MSIDTokenCacheKey keyForAccessTokenWithAuthority:tokenInCache.authority
-                                                                                      clientId:tokenInCache.clientId
-                                                                                        scopes:tokenInCache.scopes
-                                                                                        userId:account.userIdentifier];
-            
-            if (![self removeTokenWithKey:keyToDelete context:context error:error])
-            {
-                return NO;
-            }
+            return NO;
```

Comment:
does this mean that we will not save the token if we could not find intersected tokens?

Reply:
deleteAllAccessTokensWithIntersectingScopes should return NO, if deletion failed or if query failed (returned nil). This would happen with keychain errors and I don't think we should proceed. Let me know if I'm missing something.

Reply:
yes, after double check, you are right :)